src/preprocessing/passes/daneshvar.cpp

```diff
@@ -0,0 +1,817 @@
+ * Top contributors (to current version):
+ *   Daneshvar Amrollahi
+ *
+ * This file is part of the cvc5 project.
+ *
+ * Copyright (c) 2009-2023 by the authors listed in the file AUTHORS
+ * in the top-level source directory and their institutional affiliations.
+ * All rights reserved.  See the file COPYING in the top-level source
+ * directory for licensing information.
+ * ****************************************************************************
+ *
+ * The strings eager preprocess utility.
+ */
+
+#include "preprocessing/passes/daneshvar.h"
+
+#include "preprocessing/assertion_pipeline.h"
+#include "theory/rewriter.h"
+#include "theory/strings/theory_strings_preprocess.h"
+#include "expr/node_manager.h"
+#include "expr/node_manager_attributes.h"
+#include "expr/sort_to_term.h"
+#include "util/string.h"
+#include "expr/node_converter.h"
+#include "util/statistics_registry.h"
+#include "preprocessing/preprocessing_pass_context.h" 
+
+#include <map>
+#include <stack>
+
+using namespace cvc5::internal::theory;
+using namespace cvc5::internal::kind;
+
+
+namespace cvc5::internal {
+namespace preprocessing {
+namespace passes {
+
+Daneshvar::Daneshvar(PreprocessingPassContext* preprocContext)
+    : PreprocessingPass(preprocContext, "daneshvar"),
+    d_statistics(statisticsRegistry())
+    {};
+
+void dfs(const Node& n, std::string& encoding, std::vector<std::string>& varNames)
+{
+    if (n.isVar())
+    {
+        std::string varname = n.toString();
+        encoding += "V;";
+        varNames.push_back(varname);
+        return;
+    }
+    if (n.isConst())
+    {
+        encoding += n.toString() + ";";
+        return;
+    }
+
+    cvc5::internal::Kind k = static_cast<cvc5::internal::Kind>(n.getKind());
+    std::string boz = cvc5::internal::kind::toString(k);
+    encoding += boz + ";";
+    for (size_t i = 0; i < n.getNumChildren(); i++)
+    {
+        dfs(n[i], encoding, varNames);
+    }
+}
+
+
+
+
+int getRole(std::string var, NodeInfo n)
+{
+    if (n.role.find(var) != n.role.end())
+    {
+        return n.role[var];
+    }
+    return 0;
+}
+
+NodeInfo getNodeInfo(const Node& n, unsigned ecId_ass, unsigned ecId_op)
+{
+    std::string encoding = "";
+    std::vector<std::string> varNames; 
+    dfs(n, encoding, varNames);
+    encoding += "\n";
+
+    std::vector<int> pat;
+    std::map<std::string, int> varMap;
+    std::map<std::string, int> role;
+    int id = 1;
+    for (size_t i = 0; i < varNames.size(); ++i)
+    {
+        if (varMap.find(varNames[i]) == varMap.end())
+        {
+            varMap[varNames[i]] = id++;
+            role[varNames[i]] = varMap[varNames[i]];
+            // std::cout << "Role of " << varNames[i] << " is " << role[varNames[i]] << std::endl;
+        }
+        pat.push_back(varMap[varNames[i]]);
+    }
+    // std::cout << "---------------------------------" << std::endl;
+    return NodeInfo(n, encoding, pat, varNames, role, ecId_ass, ecId_op);
```

Comment:
Instead of creating the NodeInfo down here, you could add a constructor `NodeInfo ni(n);` and directly use n.pat instead of pat (same for the other arguments). This will again avoid unnecessary copies of these data structures.  You could also make use of std::move, but the former suggestion would be simpler.